protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,853 @@
+};
+use libp2p_swarm::{
+    dial_opts::{DialOpts, PeerCondition},
+    AddressScore, DialError, IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction,
+    PollParameters,
+};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    iter,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+/// Config for the [`Behaviour`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Config {
+    /// Timeout for requests.
+    pub timeout: Duration,
+
+    // == Client Config
+    /// Delay on init before starting the fist probe.
+    pub boot_delay: Duration,
+    /// Interval in which the NAT should be tested again if max confidence was reached in a status.
+    pub refresh_interval: Duration,
+    /// Interval in which the NAT status should be re-tried if it is currently is unknown
+    /// or max confidence was not reached yet.
+    pub retry_interval: Duration,
+
+    /// Throttle period for re-using a peer as server for a dial-request.
+    pub throttle_server_period: Duration,
+    /// Use connected peers as servers for probes.
+    pub use_connected: bool,
+    /// Max confidence that can be reached in a public / private NAT status.
+    /// Note: for [`NatStatus::Unknown`] the confidence is always 0.
+    pub confidence_max: usize,
+
+    //== Server Config
+    /// Max addresses that are tried per peer.
+    pub max_peer_addresses: usize,
+    /// Max total dial requests done in `[Config::throttle_clients_period`].
+    pub throttle_clients_global_max: usize,
+    /// Max dial requests done in `[Config::throttle_clients_period`] for a peer.
+    pub throttle_clients_peer_max: usize,
+    /// Period for throttling clients requests.
+    pub throttle_clients_period: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Config {
+            timeout: Duration::from_secs(30),
+            boot_delay: Duration::from_secs(15),
+            retry_interval: Duration::from_secs(90),
+            refresh_interval: Duration::from_secs(15 * 60),
+            throttle_server_period: Duration::from_secs(90),
+            use_connected: true,
+            confidence_max: 3,
+            max_peer_addresses: 16,
+            throttle_clients_global_max: 30,
+            throttle_clients_peer_max: 3,
+            throttle_clients_period: Duration::from_secs(1),
+        }
+    }
+}
+
+/// Assumed NAT status.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NatStatus {
+    Public(Multiaddr),
+    Private,
+    Unknown,
+}
+
+impl NatStatus {
+    pub fn is_public(&self) -> bool {
+        matches!(self, NatStatus::Public(..))
+    }
+}
+
+impl From<Result<Multiaddr, ResponseError>> for NatStatus {
+    fn from(result: Result<Multiaddr, ResponseError>) -> Self {
+        match result {
+            Ok(addr) => NatStatus::Public(addr),
+            Err(ResponseError::DialError) => NatStatus::Private,
+            _ => NatStatus::Unknown,
+        }
+    }
+}
+
+/// Network Behaviour for AutoNAT.
```

Comment:
```suggestion
/// [`NetworkBehaviour`] for AutoNAT.
```
Just in case we ever rename `NetworkBehaviour` :innocent: